horovod/common/__init__.py

```diff
@@ -46,12 +47,36 @@ def check_extension(ext_name, ext_env_var, pkg_path, *args):
     ctypes.CDLL(os.path.join(os.path.dirname(__file__),
                              'mpi_lib' + get_ext_suffix()), mode=ctypes.RTLD_GLOBAL)
 
-
-def init():
+def init(comm=None):
     """A function that initializes Horovod.
-    """
-    return MPI_COMMON_LIB_CTYPES.horovod_init()
 
+    Args:
+      comm: List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
+        communicator OR the MPI communicator to use. Given communicator will be duplicated.
+        If None, Horovod will use MPI_COMM_WORLD Communicator.
+
+    """
+    if comm is None:
+        comm=[]
+
+    atexit.register(terminate, finalize=True)
+
+    if not isinstance(comm, list):
+        from mpi4py import MPI
+        if MPI._sizeof(MPI.Comm) == ctypes.sizeof(ctypes.c_int):
+            MPI_Comm = ctypes.c_int
+        else:
+            MPI_Comm = ctypes.c_void_p
+            MPI_COMMON_LIB_CTYPES.horovod_init_comm.argtypes = [MPI_Comm]
+
+        comm_obj = MPI_Comm.from_address(MPI._addressof(comm))
+        return MPI_COMMON_LIB_CTYPES.horovod_init_comm(comm_obj)
+    else:
+        comm_size = len(comm)
+        return MPI_COMMON_LIB_CTYPES.horovod_init((ctypes.c_int * comm_size)(*comm), ctypes.c_int(comm_size))
+
+def terminate(finalize=False):
```

Comment:
`terminate` sounds a bit fatalistic and like something that can only be done once.  I can see why `finalize` may be loaded name here, but perhaps `shutdown` is better?